hbase-server/src/main/java/org/apache/hadoop/hbase/ccsmap/ChunkPool.java

```diff
@@ -0,0 +1,208 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ * <p>
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hadoop.hbase.ccsmap;
+
+import java.util.Map;
+import java.util.Queue;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.atomic.AtomicLong;
+
+import com.google.common.annotations.VisibleForTesting;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.hbase.classification.InterfaceAudience;
+
+/**
+ * It mainly focus on managing reusable chunks, to make GC friendy.
+ * Also, if situation needs, this class also keep tracking non-resuable chunks.
+ */
+@InterfaceAudience.Private
+public class ChunkPool {
+
+  private static final Log LOG = LogFactory.getLog(ChunkPool.class);
+  private static final Object initLock = new Object();
+  private static volatile ChunkPool globalInstance;
+
+  private final HeapMode heapMode;
+  private final int maxChunkCount;
+  private final int chunkSize;
+  private final ChunkAllocator chunkAllocator;
+  private final AbstractHeapChunk[] chunkArray;
```

Comment:
I have no idea what this array is for, but it can provide a O(1) time search. Thinking it will be up to the stage when we proceed further.